include/nbl/builtin/hlsl/type_traits.hlsl

```diff
@@ -4,13 +4,508 @@
+  template<class T, unsigned I = 0> struct extent; (TODO)
+ 
+  // type relations
+  template<class T, class U> struct is_same; (DONE)
+  template<class Base, class Derived> struct is_base_of; (TODO)
+  template<class From, class To> struct is_convertible; (TODO)
+  template<class From, class To> struct is_nothrow_convertible; (TODO: ALIAS OF is_convertible)
+  template<class T, class U> struct is_layout_compatible; (TODO)
+  template<class Base, class Derived> struct is_pointer_interconvertible_base_of; (NOT-APPLICABLE)
+ 
+  NO FOLD EXPRESSIONS IN HLSL!
+    template<class Fn, class... ArgTypes> struct is_invocable;
+    template<class R, class Fn, class... ArgTypes> struct is_invocable_r;
+    template<class Fn, class... ArgTypes> struct is_nothrow_invocable;
+    template<class R, class Fn, class... ArgTypes> struct is_nothrow_invocable_r;
+ 
+  // const-volatile modifications
+  template<class T> struct remove_const; (TODO)
+  template<class T> struct remove_volatile; (TODO)
+  template<class T> struct remove_cv; (TODO)
+  template<class T> struct add_const; (TODO)
+  template<class T> struct add_volatile; (TODO)
+  template<class T> struct add_cv; (TODO)
+
+  // reference modifications
+  template<class T> struct remove_reference; (TODO)
+  template<class T> struct add_lvalue_reference; (TODO)
+  template<class T> struct add_rvalue_reference; (TODO)
+
+  // sign modifications
+  template<class T> struct make_signed; (TODO)
+  template<class T> struct make_unsigned; (TODO)
+ 
+  // array modifications
+  template<class T> struct remove_extent; (TODO)
+  template<class T> struct remove_all_extents; (TODO)
+
+  // pointer modifications
+  template<class T> struct remove_pointer; (TODO)
+  template<class T> struct add_pointer; (TODO)
+
+  // other transformations
+  template<class T> struct type_identity; (DONE)
+  template<class T> struct remove_cvref; (TODO)
+  template<class T> struct decay; (TODO)
+  template<bool, class T = void> struct enable_if; (NOT-APPLICABLE)
+  template<bool, class T, class F> struct conditional; (DONE)
+  template<class... T> struct common_type; (NOT-APPLICABLE)
+  template<class T, class U, template<class> class TQual, template<class> class UQual>
+    struct basic_common_reference { }; (NOT-APPLICABLE)
+  template<class... T> struct common_reference; (NOT-APPLICABLE)
+  template<class T> struct underlying_type;
+  template<class Fn, class... ArgTypes> struct invoke_result;
+  template<class T> struct unwrap_reference; (TODO)
+  template<class T> struct unwrap_ref_decay; (TODO)
+
+  NO FOLD EXPRESSIONS IN HLSL!
+    // logical operator traits
+    template<class... B> struct conjunction;
+    template<class... B> struct disjunction;
+    template<class B> struct negation;
+*/
+
+// C++ headers
+#ifndef __HLSL_VERSION
+#include <type_traits>
+#include <nbl/builtin/hlsl/cpp_compat/matrix.hlsl>
+#endif
+
+#include <nbl/builtin/hlsl/macros.h>
+
 namespace nbl
 {
 namespace hlsl
 {
+namespace type_traits
```

Comment:
hmm we should probably remove this namespace since the idea is that `nbl::hlsl::==std::`